zng/resolver/context.go

```diff
@@ -37,84 +38,162 @@ func NewContext() *Context {
 
 // Row is a structure used to organize the generic type table
 // into type-specific subcomponents by category name.
-type Row struct {
-	Category string      `json:"name"`
-	Type     interface{} `json:"type"`
+type Alias struct {
+	Name string
+	Id   int
+}
+
+const (
+	TypeDefRecord = 0x80
+	TypeDefArray  = 0x81
+	TypeDefSet    = 0x82
+	TypeDefAlias  = 0x83
+)
+
+// XXX for now, we marshal into this data structure to represent the entire
+// type table.  After we update the ZNG implementation to use typedefs, we
+// will write a context table as BZNG.
+type TypeDef struct {
+	Id      int
+	Code    int
+	Aliases []Alias
+}
+
+func setTypeDef(id int, t *zng.TypeSet) TypeDef {
+	return TypeDef{
+		Id:      id,
+		Code:    TypeDefSet,
+		Aliases: []Alias{Alias{"set", t.InnerType.Id()}},
+	}
+}
+
+func vectorTypeDef(id int, t *zng.TypeVector) TypeDef {
+	return TypeDef{
+		Id:      id,
+		Code:    TypeDefArray,
+		Aliases: []Alias{Alias{"array", t.Type.Id()}},
+	}
+}
+
+func recordTypeDef(id int, t *zng.TypeRecord) TypeDef {
+	var aliases []Alias
+	for _, col := range t.Columns {
+		aliases = append(aliases, Alias{col.Name, col.Type.Id()})
+	}
+	return TypeDef{
+		Id:      id,
+		Code:    TypeDefRecord,
+		Aliases: aliases,
+	}
+}
+
+func (c *Context) makeSetType(id int, aliases []Alias) (*zng.TypeSet, error) {
+	innerID := aliases[0].Id
+	typ, err := c.lookupType(innerID)
+	if err != nil {
+		return nil, err
+	}
+	return &zng.TypeSet{Context: c, ID: id, InnerType: typ}, nil
+}
+
+func (c *Context) makeVectorType(id int, aliases []Alias) (*zng.TypeVector, error) {
+	innerID := aliases[0].Id
+	typ, err := c.lookupType(innerID)
```

Comment:
Nit: Maybe shorten to `typ, err := c.lookupType(aliases[0].ID)` in these constructors.